### ISSUE TITLE
Token exchange doesn't work without user info

### ISSUE THREAD

**Issue by Shryne** [keycloak/keycloak#12546]
### Describe the bug

I've set up Keycloak with Azure Active Directory as an identity provider. In settings, there's the possibility to enable or disable the userinfo call of this provider. Since Azure requires a separate scope to make that call, I deactivated it. This works as intended for a browser login, however, the token exchange fails with this setting.

### Version

18.0.0

### Expected behavior

The userinfo call should be optional for token exchange, too.

### Actual behavior

Token exchange isn't possible without it and there is an error message in the cli telling, that the userinfo call wasn't possible.

### Anything else?

It works, if I turn on userinfo and use a token from Azure that allows the userinfo call.

**Comment by pedroigor** [comment]
I'm not sure if we can blindly trust a subject_token and return back an internal token. I guess the `subject_token` is an access token, right?

Can't you just use an ID token as a `subject_token`?

**Comment by pedroigor** [comment]
Yeah, to invoke user info you should use an AT. Sorry, my suggestion does not make sense and also won't work because the server requires an AT to complete this type of exchange.

I guess you are using the generic OpenID Connect broker and not the Microsoft specific, is that correct?

If so, do you know if the access token from MS is an opaque token or a JWT? If the token is opaque we are forced to call the user info endpoint to resolve the subject.

**Comment by Shryne** [comment]
> I guess you are using the generic OpenID Connect broker and not the Microsoft specific, is that correct?

Yes, because there are issues with the default configuration from Keycloak for Azure.

> If so, do you know if the access token from MS is an opaque token or a JWT? If the token is opaque we are forced to call the user info endpoint to resolve the subject.

It's a JWT, but with an addition. See [AzureAD/609](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/609):
> Hello,
How was this resolved?
I have an access token that I receive from Graph API. I am trying to validate it with the key found in the https://login.microsoftonline.com/common/discovery/keys. I am using jwt.io. But it fails with Invalid Signature. The token has nonce value. What kind of special processing is required? My goal is to send the token in headers to my api and get the token validated using jwt token validation method. Any help will be great, stuck in this since past few days.

Answer:  
> @chetanku a token from Graph that has a 'nonce' requires special processing to validate the signature

Without the nonce, it's a normal JWT.

**Comment by Shryne** [comment]
Yes, I will look into it.

**Comment by sudmis** [comment]
Is user info call supported and works from keycloak to Azure AD? I am getting "user info call failure" for External to internal token exchange. Looks like it is going to "https://login.microsoftonline.com/**common**/oauth2/authorize" , which shouldn't be the case. Am I missing any configuration? Is this scenario supported and has anyone solved this?

2024-05-01 01:22:08,877 DEBUG [org.apache.http.wire] (executor-thread-99) http-outgoing-11 << "WWW-Authenticate: Bearer realm="", authorization_uri="https://login.microsoftonline.com/common/oauth2/authorize", client_id="00000003-0000-0000-c000-000000000000"[\r][\n]"

2024-05-01 01:22:08,878 DEBUG [org.apache.http.headers] (executor-thread-99) http-outgoing-11 << HTTP/1.1 401 Unauthorized

**Comment by thomasdarimont** [comment]
@sudmis is still still an issue? As @pedroigor suggested I think we can introduce an additional configuration option like "Use userinfo for token-exchange" (default to true) to the IdP configuration. If this is set to false we could adapt the logic in https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L915 to verify the subject token (JWT access token) manually with the JWK from the `https://login.microsoftonline.com/${tenantId}/discovery/v2.0/keys` JWKS endpoint. Then we could use the `sub` from the azure access token to lookup the local existing user with the linked Keycloak account for this IdP.

**Comment by bradhead** [comment]
Hey, I am getting this exact same issue with Keycloak 24.0.5  exchanging an ID and/or Access Token issued by another Keycloak. This is not limited to Azure AAD. 

The error I get is  "invalid token". The logs say  "user info call falure". Yet, if I do a GET request in Postman to the exact same userinfo url as configured in my IDP, and pass the token I got from the IDP, it works just fine. So, what is going on that the user info call would fail?

**Comment by thomasdarimont** [comment]
I made some progress with this, but I'm currently blocked by this [issue 34682](https://github.com/keycloak/keycloak/issues/34682) in KC26 and KC25: which makes it currently impossible to properly setup token exchange policies for Identity Providers.

With a local workaround I was able to perform a exchange an EntraID IdToken to a corresponding Keycloak Access Token for the user referenced by the EntraID IdToken.

```
POST https://id.acme.test:8443/auth/realms/acme-apps/protocol/openid-connect/token
Content-Type: application/x-www-form-urlencoded

grant_type=urn:ietf:params:oauth:grant-type:token-exchange&client_id=token-exchanger&client_secret={{CLIENT_SECRET}}&subject_issuer=idp-acme-azuread&subject_token={{ENTRA_ID_TOKEN}}&requested_token_type=urn:ietf:params:oauth:token-type:access_token&subject_token_type=urn:ietf:params:oauth:token-type:id_token
```

**Comment by xsreality** [comment]
I am trying to exchange an AT received via Client Credentials flow against Azure AD so the token is not associated with any user but it has the `sub` claim with client id as the value. I hit the same issue described here when trying to exchange the ÅT with Keycloak. It tries to call the Userinfo endpoint to validate the AT and fails with invalid token "user info call failure". 

The exact failure can be identified by manually calling the userinfo endpoint. I get back the error "Access token validation failure. Invalid audience." from Azure. This is expected because I am not using the graph scope in the request so the AT is not valid for Microsoft Graph. When using the graph scope `https://graph.microsoft.com/.default`, the audience is set to `https://graph.microsoft.com` which solves the previous error but now I get a new error "Token must contain sub claim". Now this might be confusing since the `sub` claim is present with the client id value. The error is actually referring to `xms_st.sub` claim which is only present when using delegated auth (involving a user).

So all this can be avoided if Keycloak doesn't insist on calling the userinfo endpoint during token exchange. Although I am not sure if exchanging an AT obtained via Client credentials from Azure is even a valid thing to do.
